x/group/keeper/msg_server.go

```diff
@@ -700,11 +679,17 @@ func (k Keeper) doTallyAndUpdate(ctx sdk.Context, p *group.Proposal, electorate
 		return err
 	}
 
-	result, err := policy.Allow(tallyResult, electorate.TotalWeight, ctx.BlockTime().Sub(submittedAt))
+	sinceSubmission := ctx.BlockTime().Sub(p.SubmitTime) // duration passed since proposal submission.
+	result, err := policy.Allow(tallyResult, electorate.TotalWeight, sinceSubmission)
+	// If the result was final (i.e. enough votes to pass) or if the voting
+	// period ended, then we consider the proposal as final.
+	isFinal := result.Final || ctx.BlockTime().After(p.VotingPeriodEnd)
```

Comment:
vulnerability 2:
before we were pruning votes only when the `TallyResult.Final` (set by the decision policy's Allow) was true. If it was false, we were never pruning votes, even after the voting period end.

Fixing here to mark the proposal as final if the `TallyResult.Final` is true OR if the voting period end passed.